server/src/main/java/org/eclipse/lsp/cobol/core/visitor/VisitorHelper.java

```diff
@@ -231,10 +213,83 @@ public String getIntervalText(ParserRuleContext ctx) {
    * Check if the current thread was interrupted and stop the further analysis by throwing a
    * specific exception
    */
-  public void checkInterruption() {
+  void checkInterruption() {
     if (Thread.interrupted()) {
       LOG.debug("Parsing interrupted by user");
       throw new ParseCancellationException("Parsing interrupted by user");
     }
   }
+
+  /**
+   * Create a tree node from the given context
+   *
+   * @param positions map of localities
+   * @param constants predefined variable context
+   * @param nameContext context of name rule
+   * @param children children nodes
+   * @param hierarchy stream of parent variables
+   * @return list of variable nodes
+   */
+  List<Node> createVariableUsage(
+      Map<Token, Locality> positions,
+      PredefinedVariableContext constants,
+      ParserRuleContext nameContext,
+      List<Node> children,
+      Stream<ParserRuleContext> hierarchy) {
+    String dataName = getName(nameContext);
+    List<Node> result = new ArrayList<>();
+    getLocality(positions, nameContext.getStart())
+        .ifPresent(
+            locality -> {
+              if (constants.contains(dataName)) constants.addUsage(dataName, locality.toLocation());
+              else {
+                result.add(
+                    new VariableUsageNode(
+                        dataName, locality, createParentsList(positions, hierarchy)));
+              }
+            });
+    result.addAll(children);
+    return result;
+  }
+
+  /**
+   * Create a tree node from the given context
+   *
+   * @param positions map of localities
+   * @param children children nodes
+   * @param ctx to retrieve the locality range
+   * @param nodeConstructor function to create the node
+   * @return list of nodes
+   */
+  List<Node> createTreeNode(
+      Map<Token, Locality> positions,
+      List<Node> children,
+      ParserRuleContext ctx,
+      Function<Locality, Node> nodeConstructor) {
```

Comment:
You are calling this function only in one place. Do you plan to use another `nodeConstructor` in the future?